Makes use of a config-file possible along with command-line options

foxish opened the pull request:
Fixes https://github.com/kubernetes/contrib/issues/1415.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1497)

<!-- Reviewable:end -->

foxish committed:
Making it possible to consume flags from file

eparis commented on mungegithub/submit-queue/deployment/contrib/configmap.yaml:
```diff
@@ -3,41 +3,42 @@ kind: ConfigMap
-  gitrepo.enable-md-yaml: "false"
+  munge-config: |
+    # basic config options.
+    http-cache-dir: /cache/httpcache
```
Should we leave things like which, which are defined in the deployment template, as command line flags? So both defining the location of the volume mounts and telling the app where those volume mounts are can be changed together? (This is true for a couple of these flags)

apelisse replied:
I think this is a good point

foxish replied:
Agreed. We don't want someone changing one without changing the other. I'll make this change.

foxish replied:
Done.

eparis commented:
CI failed because you didn't vendor all of the deps (`godep save ./...`) 

I'm also unsure if we are allows to use `BurntSushi/toml`. I remember something about that from @bgrant0607 and it is a dep of viper.

foxish commented:
@eparis Tried again with `godep save` like you mentioned. Hopefully that should work. BurntSushi/toml is under the wtfpl, which might be why we cannot use it. I could make a fork of viper that does not use it, because we don't use the TOML parser anyway if necessary.

foxish commented:
I realize that the way I've fixed the issue right now, removing the toml parsing from the vendored import files is not the right way. Would creating a fork of viper without that package be the right step? I can see that there was an effort to get rid of it in https://github.com/spf13/viper/commit/abafbf243b9583ac1360ae4ef680d8acaa361089 but it was reverted.

eparis commented:
I'll wait for @bgrant0607 to say +1 -1 on the state of the toml parser.

apelisse commented on mungegithub/mungegithub.go:
```diff
@@ -128,6 +129,22 @@ func main() {
+	viper.AddConfigPath("/etc/config/")
+	viper.AddConfigPath(".")
+	if err := viper.ReadInConfig(); err != nil {
+		glog.Warningf("Cannot read configuration file: %v", err)
```
So the config is not an option itself, it's a mandatory thing?

foxish replied:
It currently reads a config file of that name if present in one of those two locations, and sets the various option flags. If a config file is not present, it will use the flag values supplied via the commandline  

I guess we should make the configfile path an option, so that it is clearer _where_ we're reading values from. Good point.

foxish replied:
Done.

apelisse commented on mungegithub/submit-queue/deployment/contrib/configmap.yaml:
```diff
@@ -3,41 +3,42 @@ kind: ConfigMap
 metadata:
   name: contrib-sq-config
 data:
-  # basic config options.
```
Would it make sense to "create configmap --from-file" instead?

foxish replied:
I think that's a good idea. I don't want to change the Makefile as well in this PR, I'll check that approach in a subsequent one.

bgrant0607 commented:
That wasn't resolved by https://github.com/spf13/viper/pull/217 ?

cc @caesarxuchao

bgrant0607 commented:
Hmm. Ok, I guess not.

caesarxuchao commented:
Unfortunately https://github.com/spf13/viper/pull/217 got reverted.

foxish committed:
Updated godep and vendor directories.

foxish commented:
@eparis, addressed comments and fixed the parser issue by forking viper as of https://github.com/spf13/viper/pull/217 and using that as our dependency instead. PTAL.

lavalamp commented on mungegithub/mungegithub.go:
```diff
@@ -25,12 +25,13 @@ import (
 	"k8s.io/contrib/mungegithub/reports"
 	utilflag "k8s.io/kubernetes/pkg/util/flag"
 
+	"github.com/foxish/viper"
```
Probably shouldn't point at your own fork of this. Are you waiting for changes to get merged upstream?

foxish replied:
There is a discussion in the comments. There was a patch to replace the TOML library that was being used (as it is under the WTFPL), but it was reverted. @bgrant0607 suggested that we've waited long enough and should just fork it.

eparis replied:
Since the toml library which was reverted has now been fixed, I'm ok with the fork, but can you take the lead of trying to update upstream viper again?

foxish replied:
@eparis Sure, I can do that. However, @lavalamp made a good point that the huge number of dependencies we're adding for relatively simple reading of a config YAML might not be worth it. So, I think I might move away to reading the yaml ourselves in this case, unless there is a good reason to stick with viper.

lavalamp commented:
wow that dependency tree is huge. Is viper really doing something that takes more than a few hundred lines of code?

apelisse commented:
Reviewed 6 of 6 files at r1, 362 of 362 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1497)*
<!-- Sent from Reviewable.io -->

test-foxish commented:
recomputing cla status...

foxish commented:
Closing for now, as there is no real need for this, and spf13/viper can be used directly in a follow-up as toml library licensing issues have been resolved.
